### PR TITLE
builder: split nix-instantiate & nix-build

### LINKED ISSUE #50: Lorri watch panics of Nix failed to build due to a missing source file
@grahamc writes:
> Repro: https://github.com/tweag/lorri/tree/missing-source-file-panic-repro/repros/missing-source-file-panic
>
> We should filter out paths which don't exist, and not attempt to watch them.
>
> We should also pass through the error log messages.

### LINKED ISSUE #143: Improve error message when shell.nix is failing
having 
```
with import <nixpkgs> {};
pkgs.mkShell {
  buildInputs = [
    foobar
  ];
}
```

we run `RUST_BACKTRACE=full lorri watch`
we get:

```
Started
thread '<unnamed>' panicked at 'got more or less than one build product from logged_evaluation.nix: []', src/builder.rs:89:5
stack backtrace:
   0:     0x55b97cd7ddcf - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::hca90a656f82f7090
   1:     0x55b97cd8d517 - std::sys_common::backtrace::print::hc0f09baa317c9c13
   2:     0x55b97cd828b3 - std::panicking::default_hook::{{closure}}::h8b932dda5d5aa5fb
   3:     0x55b97cd82640 - std::panicking::default_hook::h73f0283673f071f7
   4:     0x55b97cd82faf - std::panicking::rust_panic_with_hook::hfb62954e8a2422c3
   5:     0x55b97cd82b41 - std::panicking::continue_panic_fmt::hdc6f6763fa99ff03
   6:     0x55b97cd82a8e - std::panicking::begin_panic_fmt::h652121450219d0d4
   7:     0x55b97cc34a07 - lorri::builder::run::hbb8cf3e23413aecf
   8:     0x55b97cc5667b - lorri::build_loop::BuildLoop::once::hfbc08a3947497490
   9:     0x55b97cc563af - lorri::build_loop::BuildLoop::forever::h50ddf3e10f88b5cc
  10:     0x55b97cc38b3f - std::sys_common::backtrace::__rust_begin_short_backtrace::h6430451947a1ce36
  11:     0x55b97cc2e6c8 - std::panicking::try::do_call::h383385312a0ac4e8
  12:     0x55b97cd97ab9 - __rust_maybe_catch_panic
  13:     0x55b97cc3f874 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h9362940600f79da0
  14:     0x55b97cd7b4be - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hc3a444b8803cf943
  15:     0x55b97cd8217f - std::sys_common::thread::start_thread::hb1320f6edf722579
  16:     0x55b97cd7e9a5 - std::sys::unix::thread::Thread::new::thread_start::hc0c954f6a39d0e2e
  17:     0x7f7f85405ef6 - start_thread
  18:     0x7f7f8512522e - __GI___clone
  19:                0x0 - <unknown>
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any', src/libcore/result.rs:999:5
stack backtrace:
   0:     0x55b97cd7ddcf - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::hca90a656f82f7090
   1:     0x55b97cd8d517 - std::sys_common::backtrace::print::hc0f09baa317c9c13
   2:     0x55b97cd828b3 - std::panicking::default_hook::{{closure}}::h8b932dda5d5aa5fb
   3:     0x55b97cd82640 - std::panicking::default_hook::h73f0283673f071f7
   4:     0x55b97cd82faf - std::panicking::rust_panic_with_hook::hfb62954e8a2422c3
   5:     0x55b97cd82b41 - std::panicking::continue_panic_fmt::hdc6f6763fa99ff03
   6:     0x55b97cd82a25 - rust_begin_unwind
   7:     0x55b97cdaf63c - core::panicking::panic_fmt::h95972081a9687957
   8:     0x55b97cc1f564 - core::result::unwrap_failed::h8210ba461eea6f78
   9:     0x55b97cc4a9ea - lorri::ops::watch::main::h14a6e1cd3d0c3889
  10:     0x55b97cc13e26 - lorri::main::h49ee70582fd361cb
  11:     0x55b97cc14522 - std::rt::lang_start::{{closure}}::h5d980ca3d7334c06
  12:     0x55b97cd829c2 - std::panicking::try::do_call::h58dd4f747553ff5f
  13:     0x55b97cd97ab9 - __rust_maybe_catch_panic
  14:     0x55b97cd932c8 - std::rt::lang_start_internal::hd752bd0add31c5fb
  15:     0x55b97cc144f4 - main
  16:     0x7f7f85054b8d - __libc_start_main
  17:     0x55b97cc132c9 - _start
  18:                0x0 - <unknown>
```
maybe we could to give a more user friendly message to the user ?

### PR DESCRIPTION
**Pull request by Profpatsch**
A bit of refactoring, to finally split the instantiation and build (see last commit).

Please review commit-by-commit.

- Add wrappers around `PathBuf` to distinguish drv files and final build results
- Use struct `OutputPaths` to keep results of `logged-evalatuion.nix` & move errors from use time to parse time
- remove `lorri shell`
- Fix an issue with nix output parsing
- and finally split instantiate and build

(Part of the fix for #58)

Closes #143 
Closes #50

### CODE REVIEW

File: src/nix.rs [hunk 1]
```diff
@@ -35,29 +35,32 @@
...
+use StorePath;
+
 /// Execute Nix commands using a builder-pattern abstraction.
 #[derive(Clone)]
-pub struct CallOpts {
-    input: Input,
+pub struct CallOpts<'a> {
+    input: Input<'a>,
     attribute: Option<String>,
+    result_gc_root_path: Option<PathBuf>,
```
Comment (grahamc) [inline src/nix.rs]: Right off the bat feedback: this isn't workable. We need to GC root everything Nix evaluates and builds, and enforce it at the type system level.

One improvement on the way it works on master is if a store path result from the Nix evaluator carries with it its tempdir handle, forcing the tempdir to stay alive until the Nix value is garbage collected.

**Comment by Profpatsch** [comment]
Rebased & ready for review.

File: src/nix.rs (hunk 1 above)
Comment (Profpatsch) [reply src/nix.rs]: So, what effectively happened is: Once I removed the `lorri shell` subcommand, no code whatsoever is using this stuff anymore. The only part that creates a gc root is the store path generated by `logged-evaluation.nix`, and that is handled by `roots`. I’m thinking about completely removing this from the builder for now.

Comment (Profpatsch) [reply src/nix.rs]: Okay, the problem with creating the roots after a build is that there is a small period of time (race condition) where a GC could delete the store path. Thus we must pass the gc root to the nix build.

Comment (Profpatsch) [reply src/nix.rs]: This is resolved by basing this branch on https://github.com/target/lorri/pull/131. PTAL

**Comment by Profpatsch** [comment]
I split this PR up in multiple PRs for easier review:

- Part 1 https://github.com/target/lorri/pull/133
- Part 2 https://github.com/target/lorri/pull/134
- Part 3 https://github.com/target/lorri/pull/135

And the rest of the commits are in this PR.

**Comment by Profpatsch** [comment]
Closes #121

**Comment by Profpatsch** [comment]
Added a new commit, because we forgot to handle the `BuildFailure` case in `CallOpts`.

**Comment by Profpatsch** [comment]
holy cow, I can’t believe it’s merged

**Pull request merged**